File: Sample/Sample/ViewController.m
//
//  ViewController.m
//  Sample
//
//  Created by roy on 2/1/16.
//  Copyright © 2016 Confirm.io. All rights reserved.
//

#import "ViewController.h"
#import "ResultsController.h"

#import <confirm_sdk/confirm_sdk.h>

@interface NSString(Sample)
+ (NSString*)memoryString:(int64_t)value;
@end

@interface ViewController () <ConfirmCaptureDelegate>

@property (weak, nonatomic) IBOutlet UIButton *workflowButton;
@property (weak, nonatomic) IBOutlet UIButton *workflowFacialButton;

@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UILabel *progressLabel;

@property (strong, nonatomic)	ConfirmPayload*	payload;

@end

@implementation ViewController
@synthesize statusLabel, progressLabel;
@synthesize workflowButton, workflowFacialButton;

+ (ViewController*)controller
{
	ViewController* vc = [[ViewController alloc] initWithNibName:@"ViewController" bundle:nil];
	vc.edgesForExtendedLayout = UIRectEdgeLeft | UIRectEdgeRight | UIRectEdgeBottom;

	return vc;
}

- (void)viewDidLoad
{
	[super viewDidLoad];
	self.title = @"Confirm.io Sample App";
	
	[self makeButtonNice:workflowButton];
	[self makeButtonNice:workflowFacialButton];
	
}

- (void)makeButtonNice:(UIButton*)button
{
	CALayer* layer = button.layer;
	
	layer.cornerRadius = 12;
	layer.borderColor = UIColor.darkGrayColor.CGColor;
	layer.borderWidth = 2;
	
	button.titleLabel.numberOfLines	= 0;
	button.titleLabel.textAlignment = NSTextAlignmentCenter;
}

- (void)setStatus:(NSString*)title message:(NSString*)message
{
	NSString* blurb = title;
	
	if (message)
		blurb = [blurb stringByAppendingFormat:@"\n%@", message];
	
	statusLabel.text = blurb;
}

- (void)didReceiveMemoryWarning
{
	[super didReceiveMemoryWarning];
	// Dispose of any resources that can be recreated.
}

#pragma mark - UX

- (IBAction)tappedWorkflow:(UIButton*)sender
{
	UINavigationController* nav = ConfirmCapture.singleton.confirmController;
	
	ConfirmCapture.singleton.delegate = self;
	ConfirmCapture.singleton.enableFacialMatch = NO;
	
	[self presentViewController:nav
					   animated:YES
					 completion:nil];
}

- (IBAction)tappedWorkflowFacial:(UIButton*)sender
{
	UINavigationController* nav = ConfirmCapture.singleton.confirmController;
	
	ConfirmCapture.singleton.delegate = self;
	ConfirmCapture.singleton.enableFacialMatch = YES;

	[self presentViewController:nav
					   animated:YES
					 completion:nil];
}

- (void)showResults:(IDModel*)validatedID facial:(FacialMatchResponse*)facialResponse
{
	ResultsController* 				rvc = ResultsController.controller;
	rvc.result = validatedID;
	rvc.facial = facialResponse;
	[self.navigationController pushViewController:rvc animated:YES];
}

#pragma mark - ConfirmCameraDelegate

// called when picture has been taken, image is valid object
- (void)ConfirmCaptureDidComplete:(ConfirmPayload*)payload
{
	[ConfirmSubmit.singleton submitIDCapturePayload:payload
										   onStatus:^(NSDictionary* _Nonnull info, ConfirmSubmitState state) {
											   NSString* title = info[kStatusInfoTitleKey];
											   NSString* message = info[kStatusInfoMessageKey];
											   
											   [self setStatus:title message:message];
										   }
										 onProgress:^(NSProgress* _Nonnull progress, ConfirmSubmitProgressType progressType) {
											 static NSTimeInterval 		lastProgress = 0;
											 NSTimeInterval 			rightNow = NSDate.timeIntervalSinceReferenceDate;
											 
											 if (progress.completedUnitCount == progress.totalUnitCount) {
												 progressLabel.text = @"";
											 } else if (rightNow-lastProgress > 1.0) {	// throttle progress to every second
												 NSString* 		blurb = [NSString stringWithFormat:@"%@ %@ out of %@",
																		 progressType == ConfirmSubmitUploadProgress ? @"Uploading" : @"Downloading",
																		 [NSString memoryString:progress.completedUnitCount],
																		 [NSString memoryString:progress.totalUnitCount]];
												 
												 progressLabel.text = blurb;
												 lastProgress = rightNow;
											 }
										 }
										  onSuccess:^(IDModel * _Nullable validatedID, FacialMatchResponse * _Nullable facialResponse) {
											  [self showResults:validatedID facial:facialResponse];
											  [ConfirmCapture.singleton cleanup];
										  }
											onError:^(NSError * _Nonnull error, NSString * _Nullable guid) {
												
												NSLog(@"submission error %@", error.localizedDescription);
												[ConfirmCapture.singleton cleanup];
											}
	 ];
	[self dismissViewControllerAnimated:YES completion:NULL];
}

// called when user cancels VC via button
- (void)ConfirmCaptureDidCancel
{
	[self dismissViewControllerAnimated:YES completion:NULL];
}

- (void)dealloc
{
}

@end

@implementation NSString(Sample)

+ (NSString*)memoryString:(int64_t)value
{
	double dval = value;
	
	if (dval < 1024)
		return [NSString stringWithFormat:@"%d B", (int)dval];
	else {
		dval /= 1024.0;
		if (dval < 1024)
			return [NSString stringWithFormat:@"%0.1f KB", dval];
		else {
			dval /= 1024.0;
			if (dval < 1024)
				return [NSString stringWithFormat:@"%0.1f MB", dval];
			else {
				dval /= 1024.0;
				if (dval < 1024)
					return [NSString stringWithFormat:@"%0.1f GB", dval];
				else {
					dval /= 1024.0;
					if (dval < 1024)
						return [NSString stringWithFormat:@"%0.1f TB", dval];
					
				}
				
			}
		}
	}
	return nil;
}

@end

File: SwiftSample/SwiftSample/ViewController.swift
//
//  ViewController.swift
//  SwiftSample
//
//  Created by roy on 8/3/16.
//  Copyright © 2016 confirm.io. All rights reserved.
//

import UIKit
// import confirm_sdk

class ViewController: UIViewController, ConfirmCaptureDelegate {
	
	@IBOutlet weak var checkIDButton: UIButton!
	@IBOutlet weak var checkIDSelfieButton: UIButton!
	@IBOutlet weak var statusLabel: UILabel!
	@IBOutlet weak var progressLabel: UILabel!
	
	var payload: ConfirmPayload!
	var lastProgress: TimeInterval = 0
	
	
	override func viewDidLoad() 
	{
		super.viewDidLoad()
	}
	
	@IBAction func didTapCheckID(_ sender: UIButton)
	{
		var cc: ConfirmCapture
		var nav: UINavigationController
		
		cc = ConfirmCapture.singleton()
		nav = cc.confirmController()
		
		cc.delegate = self
		cc.enableFacialMatch = false
		
		self.present(nav, animated: true, completion: nil)
		
		self.checkIDButton.isEnabled = false
		self.checkIDSelfieButton.isEnabled = false
		self.setStatusString("")
		self.setProgressString("")
	}
	
	@IBAction func didTapCheckIDSelfie(_ sender: UIButton)
	{
		var cc: ConfirmCapture
		var nav: UINavigationController
		
		cc = ConfirmCapture.singleton()
		nav = cc.confirmController()
		
		cc.delegate = self
		cc.enableFacialMatch = true
		self.present(nav, animated: true, completion: nil)
		
		self.checkIDButton.isEnabled = false
		self.checkIDSelfieButton.isEnabled = false
		self.setStatusString("")
		self.setProgressString("")
	}
	
	func setProgressString(_ value: NSString)
	{
		OperationQueue.main.addOperation { 
			self.progressLabel.text = value as String
		}
	}
	
	func setStatusString(_ value: NSString)
	{
		OperationQueue.main.addOperation { 
			self.statusLabel.text = value as String
		}
	}
	
	
	func confirmCaptureDidComplete(_ payload: ConfirmPayload)
	{
		ConfirmSubmit.singleton().submitIDCapture(payload,
		                                                 onSubmission: nil,
		                                                 onStatus: { (info: Dictionary, state: ConfirmSubmitState) in
															let blurb_: AnyObject? = info[kStatusInfoTitleKey] as! NSString
															var blurb: NSString = "";
															
															if blurb_ != nil {
																blurb = blurb_ as! NSString
																let message_: AnyObject? = info[kStatusInfoMessageKey] as! NSString
																
																if message_ != nil {
																	let message: NSString = message_ as! NSString
																	blurb = blurb.appendingFormat("\n%@", message)
																}
															}
															let guid_ : AnyObject? = info[kStatusInfoGuidKey] as! NSString;
															
															if guid_ != nil {
																let guid: NSString = guid_ as! NSString;
																blurb = NSString.localizedStringWithFormat("\nguid received:\n%@", guid)
															}
															self.setStatusString(blurb);
														},
		                                                 onProgress:{ (progress: Progress, progressType: ConfirmSubmitProgressType) in
															let rightNow: TimeInterval = CACurrentMediaTime()
															
															if progress.completedUnitCount == progress.totalUnitCount {
																self.setProgressString("Sample App")
															} else if rightNow-self.lastProgress > 1.0 {	// throttle progress to every second
																let blurb: NSString = NSString.localizedStringWithFormat("%@ %@ out of %@",
																	progressType == ConfirmSubmitUploadProgress ? "Uploading" : "Downloading",
																	self.memoryString(progress.completedUnitCount)!,
																	self.memoryString(progress.totalUnitCount)!)
																
																self.setProgressString(blurb)
																self.lastProgress = rightNow
															}
														},
		                                                 onSuccess:{ (validatedID, facialResponse) in
															self.showResults(validatedID, facialResponse:facialResponse)
															ConfirmCapture.singleton().cleanup()
															self.payload = nil
															self.checkIDButton.isEnabled = true;
															self.checkIDSelfieButton.isEnabled = true;
														}, 
		                                                 onError:{ (error, guid) in
															NSLog("submission error\n%@", error.localizedDescription)
															ConfirmCapture.singleton().cleanup()
															self.payload = nil
															self.checkIDButton.isEnabled = true;
															self.checkIDSelfieButton.isEnabled = true;
														}
		)
		
		self.finishIDCapture()
	}
	
	func confirmCaptureDidCancel()
	{
		self.finishIDCapture()
	}
	
	func finishIDCapture()
	{
		self.dismiss(animated: true, completion:nil)
	}
	
	func showResults(_ validatedID: IDModel?, facialResponse: FacialMatchResponse?)
 	{
		if validatedID != nil {
			var status: NSString = ""
			if validatedID!.didPass {
				status = "ID passed";
			} else if validatedID!.didFail {
				status = "ID failed";
			} else if validatedID!.isUnknown {
				status = "ID unknown"
			}
			let identity: IdentityModel = validatedID!.identity
			let classification: IdentityClassificationModel = identity.classification;
			let issuance: IdentityIssuanceModel = identity.issuance
			let bio: IdentityBioModel = identity.bio
			
			status = status.appendingFormat(" (%@ %@ %@)\n%@ %@", classification.state, classification.type, issuance.number,
			                                        bio.firstName, bio.lastName	);
			self.setStatusString(status)
		}
		
		if facialResponse != nil {
			var progress: String = "";
			switch facialResponse!.facialMatchValue {
				case FacialMatchValueUnknown:
					progress = "FacialMatch: unknown"
					break
				case FacialMatchValueFailed:
					progress = "FacialMatch: failed"
					break
				case FacialMatchValuePassed:
					progress = "FacialMatch: passed"
					break
				case FacialMatchValueError:
					progress = "FacialMatch: error"
					break
				default: break
			}
			self.setProgressString(progress as NSString)
		}
	}
	
	func memoryString(_ value: Int64) -> NSString?
	{
		var dval: Double = Double(value)
		
		if dval < 1024 {
			return NSString.localizedStringWithFormat("%d B", dval)
		} else {
			dval /= 1024.0
			if dval < 1024 {
				return NSString.localizedStringWithFormat("%0.1f KB", dval)
			} else {
				dval /= 1024.0
				if dval < 1024 {
					return NSString.localizedStringWithFormat("%0.1f MB", dval)
				} else {
					dval /= 1024.0
					if dval < 1024 {
						return NSString.localizedStringWithFormat("%0.1f GB", dval)
					} else {
						dval /= 1024.0
						if dval < 1024 {
							return NSString.localizedStringWithFormat("%0.1f TB", dval)
						}
					}
				}
			}
		}
		return nil
	}
}

